File: bin/model_funcs/eflux.py
# python implementation of eflux.f

from bin.Field import Field, max
from bin.Grid import Grid
from bin.Workspace import Workspace
import numpy as np

def eflux(model, ws, state, dw):
    """
    calculate convective fluxes 
    
    Args:
        model (NavierStokes): physics model
        workspace (Workspace): the relevant fields
        state (Field): containing the density, x-momentum, y-momentum, and energy
        dw (Field): to store new residuals after completing fluxes
        
    """
    
    # take a workspace ws and calculate convective fluxes

    pad = model.padding
    w = state # state vector
    porJ = ws.get_field('porJ', model.className) # porosity
    p = ws.get_field('p', model.className)   # pressure
    n = Field.dim(state) # number of quantities being convected

    [nx, ny] = ws.field_size()
    [nxp, nyp] = [pad+nx+pad, pad+ny+pad]
    ip = pad
    jp = pad
    il = nx+1
    jl = ny+1
    ie = nx+2
    je = ny+2
    ib = nx+3
    jb = ny+3

    # flux array
    fs = Field((ib+1, jb+1, n))

    # i direction
    dx = ws.edge_normals(0)
    dyx = dx[:,:,0]
    dyy = dx[:,:,1]
    p_avg = p[1:ie, jp:je] + p[ip:ib, jp:je]

    # flux operator
    qsp = (dyy*w[ip:ib, jp:je, 1] - dyx*w[ip:ib, jp:je, 2]) / w[ip:ib, jp:je, 0]
    qsm = (dyy*w[ 1:ie, jp:je, 1] - dyx*w[ 1:ie, jp:je, 2]) / w[ 1:ie, jp:je, 0]

    # add up on faces
    fs[1:ie, jp:je, 0] = qsp * w[ip:ib, jp:je, 0]                   + qsm*w[1:ie, jp:je, 0] # density
    fs[1:ie, jp:je, 1] = qsp * w[ip:ib, jp:je, 1]                   + qsm*w[1:ie, jp:je, 1]  + dyy*p_avg # x - momentum
    fs[1:ie, jp:je, 2] = qsp * w[ip:ib, jp:je, 2]                   + qsm*w[1:ie, jp:je, 2]  - dyx*p_avg # y - momentum
    fs[1:ie, jp:je, 3] = qsp *(w[ip:ib, jp:je, 3] + p[ip:ib, jp:je]) + qsm*(w[1:ie, jp:je, 3] + p[1:ie, jp:je]) # energy

    # now add everything up
    dw[ip:ie, jp:je, :] = fs[ip:ie, jp:je, :] - fs[1:il, jp:je, :]


    # j direction
    dx = ws.edge_normals(1)
    dxx = dx[:,:,0]
    dxy = dx[:,:,1]
    p_avg = p[ip:ie, 1:je] + p[ip:ie, jp:jb]

    # flux operator
    qsp = porJ * (dxx*w[ip:ie, jp:jb, 1] - dxy*w[ip:ie, jp:jb, 2]) / w[ip:ie, jp:jb, 0]
    qsm = porJ * (dxx*w[ip:ie,  1:je, 1] - dxy*w[ip:ie,  1:je, 2]) / w[ip:ie,  1:je, 0]

    # add up on faces
    fs[ip:ie, 1:je, 0] = qsp*w[ip:ie, jp:jb, 0]                     + qsm*w[ip:ie, 1:je, 0] # density
    fs[ip:ie, 1:je, 1] = qsp*w[ip:ie, jp:jb, 1]                     + qsm*w[ip:ie, 1:je, 1]  + dxy*p_avg # x - momentum
    fs[ip:ie, 1:je, 2] = qsp*w[ip:ie, jp:jb, 2]                     + qsm*w[ip:ie, 1:je, 2]  - dxx*p_avg # y - momentum
    fs[ip:ie, 1:je, 3] = qsp*(w[ip:ie,jp:jb, 3] + p[ip:ie, jp:jb]) + qsm*(w[ip:ie, 1:je, 3] + p[ip:ie, 1:je]) # energy

    # now add everything up for j direction
    dw[ip:ie, jp:je, :] += fs[ip:ie, jp:je, :] - fs[ip:ie, 1:jl, :]

    stop = 0

File: bin/model_funcs/eflux.f
      subroutine eflux(w,dw,p,x,porj,il,jl,ib,jb)
c
c     ******************************************************************
c     *                                                                *
c     *   euler flux                                                   *
c     *                                                                *
c     ******************************************************************
c
c     w(i,j,1)  = density
c     w(i,j,2)  = momentum in x direction
c     w(i,j,3)  = momentum in y direction
c     w(i,j,4)  = total energy
c
c     ******************************************************************
c
c      use dims
c
c     ******************************************************************
c
c      use flo_var
c      use mesh_var
c      use solv_var
c
c     ******************************************************************
c
      implicit none

c     inputs
      integer, intent(inout) :: il,jl,ib,jb

c     everything is zero indexed except for porosity
      real(8), intent(in), dimension(0:ib,0:jb,4) :: w
      real(8), intent(in), dimension(0:ib,0:jb,2) :: x
      real(8), intent(in), dimension(0:ib,0:jb)   :: p
      real(8), intent(in), dimension(:,:)         :: porj


      real(8), intent(inout), dimension(0:ib,0:jb,4)  :: dw

      
      
      
c
c     ******************************************************************
c
c     local variables
c
c     ******************************************************************
c
      integer :: i,j,n
c
c     ******************************************************************
c
      real(8)     :: xx,yx,xy,yy
      real(8)     :: qsp,qsm,pa
c
c     ******************************************************************
c
      real(8), dimension(il+2,jl+2,4) :: fs
c
c     ******************************************************************
c
c     flux in the i direction
c
      do j=2,jl
      do i=1,il
         xy        = x(i,j,1)  -x(i,j-1,1)
         yy        = x(i,j,2)  -x(i,j-1,2)
         pa        = p(i+1,j)  +p(i,j)
         qsp       = (yy*w(i+1,j,2)  -xy*w(i+1,j,3))/w(i+1,j,1)
         qsm       = (yy*w(i,j,2)  -xy*w(i,j,3))/w(i,j,1)
         fs(i,j,1) = qsp*w(i+1,j,1)  +qsm*w(i,j,1)
         fs(i,j,2) = qsp*w(i+1,j,2)  +qsm*w(i,j,2)  +yy*pa
         fs(i,j,3) = qsp*w(i+1,j,3)  +qsm*w(i,j,3)  -xy*pa
         fs(i,j,4) = qsp*(w(i+1,j,4)  +p(i+1,j))
     .              +qsm*(w(i,j,4)    +p(i,j))
      end do
      end do
c
c     accumulate the flux in the i direction
c
      do n=1,4
         do j=2,jl
         do i=2,il
            dw(i,j,n) = fs(i,j,n)  -fs(i-1,j,n)
         end do
         end do
      end do
c
c     flux in the j direction
c
      do j=1,jl
      do i=2,il
         xx        = x(i,j,1)  -x(i-1,j,1)
         yx        = x(i,j,2)  -x(i-1,j,2)
         pa        = p(i,j+1)  +p(i,j)
         qsp       = porj(i,j)*(xx*w(i,j+1,3)  -yx*w(i,j+1,2))/
     .                          w(i,j+1,1)
         qsm       = porj(i,j)*(xx*w(i,j,3)  -yx*w(i,j,2))/
     .                          w(i,j,1)
         fs(i,j,1) = qsp*w(i,j+1,1)  +qsm*w(i,j,1)
         fs(i,j,2) = qsp*w(i,j+1,2)  +qsm*w(i,j,2)  -yx*pa
         fs(i,j,3) = qsp*w(i,j+1,3)  +qsm*w(i,j,3)  +xx*pa
         fs(i,j,4) = qsp*(w(i,j+1,4)  +p(i,j+1))
     .              +qsm*(w(i,j,4)    +p(i,j))
      end do
      end do
c
c     accumulate the flux in the j direction
c
      do n=1,4
         do j=2,jl
         do i=2,il
            dw(i,j,n) = dw(i,j,n)  +fs(i,j,n)  -fs(i,j-1,n)
         end do
         end do
      end do

      return

      end
